catroid/src/main/java/org/catrobat/catroid/embroidery/EmbroideryManager.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.embroidery;
+
+import android.content.res.Resources;
+import android.graphics.PointF;
+import android.util.DisplayMetrics;
+
+import org.catrobat.catroid.ProjectManager;
+import org.catrobat.catroid.common.BrickValues;
+import org.catrobat.catroid.content.XmlHeader;
+
+import java.util.ArrayList;
+
+public final class EmbroideryManager {
+	private static ArrayList<PointF> stitchPoints = new ArrayList<>();
+	private static float stitchSize = BrickValues.STITCH_SIZE * calculateScreenRatio();
+	private static int lastIndex = 0;
+
+	private EmbroideryManager() {
+
+	}
+
+	private static float calculateScreenRatio() {
+		DisplayMetrics metrics = Resources.getSystem().getDisplayMetrics();
+		XmlHeader header = ProjectManager.getInstance().getCurrentProject().getXmlHeader();
+		float deviceDiagonalPixel = (float) Math.sqrt(Math.pow(metrics.widthPixels, 2) + Math.pow(metrics.heightPixels, 2));
+		float creatorDiagonalPixel = (float) Math.sqrt(Math.pow(header.getVirtualScreenWidth(), 2)
+				+ Math.pow(header.getVirtualScreenHeight(), 2));
+		return creatorDiagonalPixel / deviceDiagonalPixel;
+	}
+
+	public static void addStitchPoint(PointF point) {
+		if (stitchPoints.isEmpty() || isNewStitchPoint(point)) {
+			stitchPoints.add(point);
+			lastIndex = stitchPoints.size() - 1;
+		}
+	}
+
+	private static Boolean isNewStitchPoint(PointF point) {
+		if (point.x != stitchPoints.get(lastIndex).x || point.y != stitchPoints.get(lastIndex).y) {
```

Comment:
or just simply:

return stitchPoints.get(stitchPoints.size() - 1).equals(point)

besides that, maybe remove lastIndex, because this is really unnecessary and errorprone to manually keep the size of that list in a seperate variable.

Reply:
meant
return !stitchPoints.get(stitchPoints.size() - 1).equals(point)